sql/2025/fonts/design/fonts_family_by_script.sql

```diff
@@ -0,0 +1,49 @@
+-- Section: Design
+-- Question: Which families are used broken down by script?
+-- Normalization: Requests (parsed only)
+
+-- INCLUDE https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/{year}/fonts/common.sql
+
+WITH
+requests AS (
+  SELECT
+    client,
+    SCRIPTS(payload) AS scripts,
+    FAMILY(payload) AS family,
+    COUNT(0) OVER (PARTITION BY client) AS total
+  FROM
+    `httparchive.crawl.requests`
+  WHERE
+    date = @date AND
+    type = 'font' AND
+    is_root_page AND
+    IS_PARSED(payload) AND
+    -- Without the subsampling, the query throws “UDF out of memory.” To work
+    -- around this, a random half is removed, and the counts are doubled.
+    RAND() < 0.5
```

Comment:
This seems really suspicious! How do you know the randomness is even and so you should double?

Can't you use `rank` instead to restrict this to the top million? Or run desktop and mobile separately?

Reply:
Subsampling using a random-number generator only fails when the sample size is small, which is known as the law of small numbers. Overall, the subsample is guaranteed to be representative of the population (the entire table). However, we group by two columns, which zooms in on subpopulations that can be very small. If you just scroll through the results, you will find a lot of such cases, and all of them are not trustworthy:

https://docs.google.com/spreadsheets/d/1otdu4p_CCI70B4FVzw6k02frStsPMrQoFu7jUim_0Bg/edit?gid=74546850#gid=74546850

To avoid misinterpretation, I will try to split it by device so that the entire table is visible in the results.

Reply:
Unwrapping custom functions (if possible) may help with performance.

Reply:
Hm, I got rid of JSON_EXTRACT_STRING_ARRAY, as Max suggested, and it worked. But I guess timing also plays a role. Maybe this morning, there were more resources available.